### ISSUE TITLE
[HOLD for payment 2024-06-13] [$1000] Chat - Thread - Keyboard does not open when click of reply in thread

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#19558]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any chat
2. Click on reply in thread

## Expected Result:
Keyboard opens when click of reply in thread

## Actual Result:
Keyboard does not open when click of reply in thread

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.17.2

**Reproducible in staging?:** yes

**Reproducible in production?:** yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/c60b7e28-cc55-49e5-bb8b-7c8a92480e21

https://github.com/Expensify/App/assets/93399543/6951e8ca-204b-42f0-ab2e-1a81cb50685e

**Expensify/Expensify Issue URL:**

**Issue reported by:** Manan Gadhiya (non GH)

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684466830344199

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bb2845ef5dc4bb85</li>
        <li>Upwork Job ID: 1663649941210693632</li>
        <li>Last Price Increase: 2023-06-20</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @adelekennedy</details>

**Comment by bernhardoj** [comment]
Looks like we don't auto focus when we open a chat anymore.

**Comment by tjferriss** [comment]
Whoops, I forgot to throw the `External` label on this one.

**Comment by rvenky125** [comment]
I willing to try it out. But, how can I run it in my Windows System. The .sh files are running in windows machine which are executing by running yarn run android.

**Comment by rvenky125** [comment]
Contributor details
Your Expensify account email: venkypaithireddy@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~014e86511f26b3d333

**Comment by mollfpr** [comment]
I am reviewing now...

**Comment by mollfpr** [comment]
>I solved this issue using a timeout in ReportActionCompose's ComponentDidMount. It works smoothly.

@Nodebrute We can avoid the `setTimeout` workaround for now.

@rvenky125 Isn't that will flashing the input border will blur then focus the input again?

**Comment by rvenky125** [comment]
The flash is not noticeable if we `focus` it quickly after `blur` it. I mean there should be no `setTimeout`.
```
   if (!isKeyboardVisible) {
       inputRef.current.blur()
       inputRef.current.focus()
   }
```

**Comment by mollfpr** [comment]
@Nodebrute @rvenky125 I'm still unsure if your solution is right. We have autofocus upon creating a new chat, and it's opening the keyboard fine.

https://github.com/Expensify/App/assets/25520267/c44c590b-a2db-40b7-84fd-6a1bc6b81486

**Comment by rvenky125** [comment]
I unable to run the app in my windows environment. The scripts folder contains .sh files and they are not executable on windows 😔

**Comment by mollfpr** [comment]
@rvenky125 Sorry, I don't know how to set up the App on Windows, but probably someone in the Slack channel has done it before.

**Comment by rvenky125** [comment]
> ## Proposal
> ### Please re-state the problem that we are trying to solve in this issue.
> The keyboard does not open when clicking on reply in the thread
> 
> ### What is the root cause of that problem?
> In some situations the keyboard may not open even if the text input is focused. So you can fix this issue by blurring and focusing again the text input.
> 
> ### What changes do you think we should make in order to solve the problem?
> If the keyboard still not opens while focusing the text input then you can blur the text input by calling `textInputRef.current.blur()` method. After that focus the text input again by calling `textInputRef.current.blur()`. You can check for the keyboard whether it has opened or not by adding listener to the keyboard.
> 
> ```
> import { useEffect, useState } from "react";
> import { Keyboard } from "react-native";
> 
> const useIsKeyboardVisible = () => {
> const [isKeyboardVisible, setKeyboardVisible] = useState(false);
> 
> useEffect(() => {
>   const keyboardDidShowListener = Keyboard.addListener(
>     'keyboardDidShow',
>     () => {
>       setKeyboardVisible(true); // or some other action
>     },
>   );
>   const keyboardDidHideListener = Keyboard.addListener(
>     'keyboardDidHide',
>     () => {
>       // console.log('Hide keyboard');
>       setKeyboardVisible(false); // or some other action
>     },
>   );
> 
>   return () => {
>     keyboardDidHideListener.remove();
>     keyboardDidShowListener.remove();
>   };
> }, []);
> 
>  return isKeyboardVisible;
>  };
> 
>  export default useIsKeyboardVisible;
> ```
> 
> Use this function inside the chat screen as `const isKeyboardVisible = useIsKeyboardVisible()`. you can blur the text input, focus it again if `isKeyboardVisible` is false with some delay using `setTimeout`
> 
> ### What alternative solutions did you explore? (Optional)

> The flash is not noticeable if we `focus` it quickly after `blur` it. I mean there should be no `setTimeout`.
> 
> ```
>    if (!isKeyboardVisible) {
>        inputRef.current.blur()
>        inputRef.current.focus()
>    }
> ```

Please try this, it'll solve the issue if the keyboard is not opened for any reason. You can use it inside `componentDidMount`.

**Comment by mollfpr** [comment]
@rvenky125 yes that will solve the issue, but the root cause is not addressed well. We don’t have the keyboard issue with opening a new chat.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Keyboard does not show when we press reply in thread from context menu in native.

### What is the root cause of that problem?
First of all, we can notice that the context menu is late to hide when we press Reply in thread. We have another issue to handle that here https://github.com/Expensify/App/issues/19776. But that alone won't solve this issue. I notice the auto focus thing will always have an issue when we try to focus it after a modal is closed. The example we have right now is new chat/group page in iOS. You will notice the auto focus does not work. We previously didn't have this issue because in the previous version, the text input will be shown after the transition is end, but it is not the case anymore.

### What changes do you think we should make in order to solve the problem?

1. Wait for https://github.com/Expensify/App/issues/19776 is solved
2. Auto focus the composer in `onEntryTransitionEnd`
`onEntryTransitionEnd` props available on ScreenWrapper. There are bunch of pages that use this to do the auto focus, example.
https://github.com/Expensify/App/blob/34ec27c6507c21a094536d81948c9486ad845804/src/pages/tasks/NewTaskDetailsPage.js#L75

(This is why the auto focus issue does not happen on new task page).

For our case, we want to auto focus composer, but we don't have access to its ref and we need to call the focus from ReportScreen. Luckily we have `ReportActionComposeFocusManager.focus` that we can call everywhere.

```
<ScreenWrapper 
    onEntryTransitionEnd={() => ReportActionComposeFocusManager.focus(true)}
>
```

`ReportActionComposeFocusManager.focus` will call the callback set in the composer here:
https://github.com/Expensify/App/blob/7ac94087a22a9cfd82f9a85abf3abb7278c6ba8a/src/pages/home/report/ReportActionCompose.js#L226C5-L235
We will add a new param `forceFocus` that will bypass the condition we have. `willBlurTextInputOnTapOutside` is always false for native.

**Comment by mollfpr** [comment]
I need more time to review this.

**Comment by tjferriss** [comment]
@mollfpr any chance you can review this today or tomorrow?

**Comment by mollfpr** [comment]
@tjferriss I just tried @bernhardoj suggestion, and it is because the context menu delay on hiding is mainly the root cause. If https://github.com/Expensify/App/issues/19776 solves the delay on hiding the context menu, then we can continue with @bernhardoj [proposal](https://github.com/Expensify/App/issues/19558#issuecomment-1582531463).

**Comment by tjferriss** [comment]
Ok so are we essentially holding this issue on https://github.com/Expensify/App/issues/19776?

**Comment by mollfpr** [comment]
@tjferriss Yes and it seems the proposal is reviewed there.

**Comment by tjferriss** [comment]
Still holding on #19776

**Comment by mollfpr** [comment]
Still holding

**Comment by tjferriss** [comment]
It looks like the PR we're holding on has been merged, but still needs to hit production https://github.com/Expensify/App/pull/21479

**Comment by mollfpr** [comment]
@tjferriss The PR is in production. We can off-hold this, and I will try if it's still an issue.

**Comment by mollfpr** [comment]
@bernhardoj I just tested the current `main` branch, and now the composer is focused on replying in the thread, but the keyboard is still not opened. Is your proposal still valid for that issue?

**Comment by mollfpr** [comment]
> So, I think we should find another solution that doesn't involve any delay, whether it's setTimeout or onEntryTransitionEnd.

@bernhardoj I agree with this.

What I think is weird is that it's working on opening a new chat, what's the difference between open a new chat and replying in a thread in this case?

**Comment by tjferriss** [comment]
Still on hold, MelvinBot.

**Comment by tjferriss** [comment]
Still on hold, MelvinBot.

**Comment by johnmlee101** [comment]
no update!

**Comment by tjferriss** [comment]
Agreed, I've taken it off hold. @mollfpr what do you think of the [updated proposal](https://github.com/Expensify/App/issues/19558#issuecomment-1582531463)?

**Comment by tjferriss** [comment]
Waiting for proposals to be reviewed

**Comment by mollfpr** [comment]
Sorry for the delay; catching up on the upstream PR. I agree the solution is to add the `onDismiss` to the Android.

**Comment by tjferriss** [comment]
@mollfpr any updates?

**Comment by tjferriss** [comment]
@mollfpr any updates on reviewing the proposals?

**Comment by mollfpr** [comment]
Sorry, I miss this issue. I'll check again if it's related to the `onModalHide` of react native modal.

**Comment by tjferriss** [comment]
Thanks, @mollfpr. Have you had a chance to look into this one?

**Comment by tjferriss** [comment]
@mollfpr bumping this once more ahead of the weekend.

**Comment by mollfpr** [comment]
Sorry, I got cold yesterday. I'll give an update today 🙏

**Comment by tjferriss** [comment]
Sorry to hear, hope you're feeling better. Please share an update when you have the time.

**Comment by mollfpr** [comment]
I'm back!

Great job @bernhardoj, on the PR upstream! I see some movements there.

I think @bernhardoj is committing to the new feature on the RN upstream. This part will significantly help us clean up the workaround regarding the modal hide issue. But I don't think we will get PR upstream shipped soon since @bernhardoj hasn't received feedback from the RN maintainer.

@tjferriss Should we assign the issue to them?

**Comment by tjferriss** [comment]
Assign to @bernhardoj? That works for me.

**Comment by tjferriss** [comment]
Still on hold

**Comment by mollfpr** [comment]
Sill on hold

**Comment by tjferriss** [comment]
I just tested this on iOS and the keyboard opened reliably. It looks like this issue has been resolved.

**Comment by jliexpensify** [comment]
Re-opening so we can look into an Android fix.

**Comment by tjferriss** [comment]
Got it, sounds like a plan. I'm going to put this on hold.

**Comment by tjferriss** [comment]
We're still waiting on an [upstream PR](https://github.com/Expensify/App/issues/23959#issuecomment-1686506472)

**Comment by tjferriss** [comment]
My personal opinion would be to hold on the upstream PR and avoid workarounds. @mollfpr @johnmlee101 what do you think?

**Comment by johnmlee101** [comment]
Yeah, let's see what resources we have to get more eyes from the RN on those PRs. Ideally we should avoid workarounds unless its a final resort

**Comment by mollfpr** [comment]
Not overdue, but it still holds for the PR upstream.

> My personal opinion would be to hold on the upstream PR and avoid workarounds.

I agree with this.

**Comment by tjferriss** [comment]
We're still on hold

**Comment by tjferriss** [comment]
Considering the hold, should we push this out to a weekly?

**Comment by tjferriss** [comment]
Can someone with an Android confirm whether this is still reproducible?

**Comment by tjferriss** [comment]
Still on hold

**Comment by tjferriss** [comment]
Still on hold

**Comment by tjferriss** [comment]
Still on hold

**Comment by tjferriss** [comment]
Still on hold

**Comment by tjferriss** [comment]
Still on hold

**Comment by johnmlee101** [comment]
Should we close this out if we can't reproduce it?

**Comment by bernhardoj** [comment]
@johnmlee101 I can still reproduce this. 

https://github.com/Expensify/App/assets/50919443/27b418b5-7c7b-4176-841f-c5753eb4bc1d


The upstream [PR](https://www.github.com/facebook/react-native/pull/42014) is almost there. I think we can make this monthly because we will also need to wait for it to be released on a newer version of RN.

**Comment by johnmlee101** [comment]
This is still on Hold, right? @bernhardoj

**Comment by bernhardoj** [comment]
@johnmlee101 correct

**Comment by adelekennedy** [comment]
taking this over from TJ!

**Comment by adelekennedy** [comment]
we're still holding here

**Comment by adelekennedy** [comment]
@bernhardoj is this still reproducible? I think we should keep this open if so and I've added it to vip-vsb

**Comment by bernhardoj** [comment]
It's still reproducible, but this time, the root cause is different (idk what has changed that makes the initial root cause outdated)

https://github.com/Expensify/App/assets/50919443/023f92fd-f045-41ba-b4d9-e8659d4b90b5

A bit technical, the condition to auto focus the composer is when `isEmptyChat` is true.
https://github.com/Expensify/App/blob/35b4d1b5a91f222986d9a887ecd96ea1ede0f594/src/pages/home/ReportScreen.tsx#L284

However, the condition is wrong because we check whether the report actions length is 0 or not. This happens after this [PR](https://github.com/Expensify/App/pull/30168). Previously, we check whether the report actions length is 1 or not to include the CREATED action that always exist.

<img width="606" alt="image" src="https://github.com/Expensify/App/assets/50919443/8c1d9bed-4498-4dea-a380-17a313b59ac5">

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Keyboard does not show when we press reply in thread from context menu in native.

### What is the root cause of that problem?
The condition to auto focus depends on `isEmptyChat` condition.
https://github.com/Expensify/App/blob/eef5fde8536c78726dcbcf1632839b803b7d8526/src/pages/home/report/ReportActionCompose/ComposerWithSuggestions/ComposerWithSuggestions.tsx#L286-L287

`isEmptyChat` checks whether the report actions length is 0 or not. However, this will be false most of the times because empty chat has 1 action, that is CREATED action.
https://github.com/Expensify/App/blob/eef5fde8536c78726dcbcf1632839b803b7d8526/src/pages/home/ReportScreen.tsx#L283

This happens after this https://github.com/Expensify/App/pull/30168. Previously, we check whether the report actions length is 1 or not to include the CREATED action that always exist.

So, `shouldAutoFocus` condition becomes `false`.

### What changes do you think we should make in order to solve the problem?

Update `isEmptyChat` condition to be like before,
```
const isEmptyChat = useMemo((): boolean => reportActions.length === 1, [reportActions]);
```

or we can use the `isEmptyChat` condition from `ReportUtils.shouldReportBeInOptionList`.
https://github.com/Expensify/App/blob/eef5fde8536c78726dcbcf1632839b803b7d8526/src/libs/ReportUtils.ts#L4615-L4616

**Comment by bernhardoj** [comment]
@adelekennedy since I don't see the new root cause is being fixed anywhere, I have posted a new proposal to address the new root cause and provided a new solution. I think we can re-review the new root cause and solution here.

cc: @mollfpr

**Comment by adelekennedy** [comment]
@mollfpr please review the new solution - I think we should take this off of hold and make this external - unless this needs BE support @johnmlee101?

**Comment by adelekennedy** [comment]
@johnmlee101 @mollfpr can either of you weigh in on the related issue [here](https://www.github.com/Expensify/App/issues/40347) that @bernhardoj has found? should we take this off hold to solve that issue?

**Comment by mollfpr** [comment]
Sorry for the delay 🙏

I tried the solution, and it's working great on fixing the issue, but the composer is focused on starting a new chat. Currently, it has not. So I don't know if this would be expected.

**Comment by bernhardoj** [comment]
@mollfpr that would be expected since it's an empty chat. When we reply in a new thread, it's an empty chat too and we should autofocus. The current behavior on `main` is the broken one.

The autofocus when chat is empty was added in this old [PR](https://github.com/Expensify/App/pull/4921) which was based on this [comment](https://github.com/Expensify/App/issues/4678#issuecomment-904413983).

**Comment by adelekennedy** [comment]
@mollfpr @bernhardoj it sounds like we're getting somewhere with the discussion above. @mollfpr is the explanation [here](https://github.com/Expensify/App/issues/19558#issuecomment-2071254758) satisfactory?

**Comment by adelekennedy** [comment]
Bump @mollfpr

**Comment by mollfpr** [comment]
Thanks @bernhardoj!

@adelekennedy @bernhardoj I think we can continue with the PR now. @johnmlee101 Do you have a question about the [proposal](https://github.com/Expensify/App/issues/19558#issuecomment-2058467702)?

**Comment by bernhardoj** [comment]
Should I open the PR now or wait for @johnmlee101?

**Comment by johnmlee101** [comment]
I think that's fine, let's go with the same shared logic as from `ReportUtils.shouldReportBeInOptionList`.

**Comment by bernhardoj** [comment]
PR is ready

cc: @mollfpr

**Comment by adelekennedy** [comment]
I think we're just waiting for John's review [here](https://github.com/Expensify/App/pull/41402#issuecomment-2110945644)

**Comment by mollfpr** [comment]
> [@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

No offending PR.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

The regression step should be enough.

>  [@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

1. Open a chat with an existing message
2. Verify the composer (if android/iOS/mWeb) isn't focused (if web/desktop) is focused
3. Send a message
4. Reply in a thread of the message
5. Verify the composer is focused and a keyboard is shown for mobile devices (except mWeb Safari because of this https://github.com/Expensify/App/issues/10414)
6. 👍 or 👎 

@adelekennedy Could you give the payment summary? Thank you!

**Comment by adelekennedy** [comment]
Ugh - Upwork won't let me reuse the original job posting so I've just created a [new job](https://www.upwork.com/jobs/~01c296b470cd77272a) and sent an offer to @bernhardoj 

**Payouts due:**

- [ ] Contributor: $1000 @bernhardoj (Upwork)
- [ ] Contributor+: $1000 @mollfpr (NewDot)

Upwork job is [here](https://www.upwork.com/ab/applicants/1801644253924164666/offers).

**Comment by bernhardoj** [comment]
Accepted

**Comment by JmillsExpensify** [comment]
$1,000 approved for @mollfpr
